DEMO: service: Add support hard coding iface name to MAC

cathay4t opened the pull request:
Some OS upgrade might cause interface name changes. To solve that:

 1. Before upgrade,
    * mkdir /etc/nmstate/pin_iface_name/
    * nmstatectl show -k | sudo tee /etc/nmstate/pin_iface_name/pin.yml
    * sudo systemctl enable nmstate

 2. Reboot to new version of OS and let nmstate.service run.
 3. Reboot again.

The `nmstate.service` will create file like if it found any interface
changed name according to stored state in /etc/nmstate/pin_iface_name/pin.yml

    /etc/systemd/network/98-ens9.link

With content:

```
[Match]
MACAddress=01:23:45:67:89:ab

[Link]
Name=ens9
```

After the second reboot, the systemd-udevd will use the old name ens9.

You may refer to the screen recording for detail: https://people.redhat.com/fge/demo/nmstate_pin_iface_name.ogv

cgwalters commented on rust/src/cli/service.rs:
```diff
@@ -99,3 +111,86 @@ fn relocate_file(file_path: &Path) -> Result<(), CliError> {
+
+fn pin_iface_name(cfg_dir: &Path) -> Result<(), CliError> {
+    let file_path = cfg_dir.join(PIN_STATE_FILENAME);
+    let mut fd = std::fs::File::open(&file_path)?;
+    let mut content = String::new();
+    fd.read_to_string(&mut content)?;
```
You almost always want to use a `BufReader` when reading from files.  This is actually just shorter and faster as e.g.:

```
    let content = std::fs::read_to_string(&file_path)?;
```

cgwalters commented on rust/src/cli/service.rs:
```diff
@@ -99,3 +111,86 @@ fn relocate_file(file_path: &Path) -> Result<(), CliError> {
+    }
+
+    let content =
+        format!("[Match]\nMACAddress={mac}\n\n[Link]\nName={iface_name}\n");
```
Let's add a comment in here like `# Generated by nmstate pin-nic-name`

cgwalters commented on rust/src/cli/service.rs:
```diff
@@ -99,3 +111,86 @@ fn relocate_file(file_path: &Path) -> Result<(), CliError> {
+    // 99-default.link
+    let file_path = link_dir.join(format!("98-{iface_name}.link"));
+
+    let mut fd = std::fs::OpenOptions::new()
+        .write(true)
+        .truncate(true)
+        .create(true)
+        .open(&file_path)
+        .map_err(|e| {
+            CliError::from(format!(
+                "Failed to store captured states to file {}: {e}",
+                file_path.display()
+            ))
+        })?;
+    fd.write_all(content.as_bytes())?;
```
I think this is all just:
```
    std::fs::write(&file_path, content.as_bytes()).map_err(|e| ...)?;
```

cathay4t replied:
it does not override, I guess. Just copy from my code bank.

cgwalters commented on rust/src/cli/service.rs:
```diff
@@ -99,3 +111,86 @@ fn relocate_file(file_path: &Path) -> Result<(), CliError> {
+    let mut fd = std::fs::File::open(&file_path)?;
+    let mut content = String::new();
+    fd.read_to_string(&mut content)?;
+    let pin_state: NetworkState = serde_yaml::from_str(&content)?;
```
Although, it's actually a bit more efficient to let serde do the reading from the file directly (not that this matters) like this:

```
    let mut r = std::fs::File::open(&file_path).map(std::io::BufReader::new);
    let pin_state: NetworkState = serde_yaml::from_reader(&mut r)?;
```

to avoid allocating an intermediate string.

cgwalters commented on rust/src/cli/service.rs:
```diff
@@ -99,3 +111,86 @@ fn relocate_file(file_path: &Path) -> Result<(), CliError> {
+            for pin_iface in pin_state
+                .interfaces
+                .iter()
+                .filter(|i| i.iface_type() == InterfaceType::Ethernet)
```
We need to think about bridge/bonds here right?

cathay4t replied:
virtual interface name never changes by kernel or OS upgrade.

cgwalters commented on rust/src/cli/service.rs:
```diff
@@ -15,6 +21,12 @@ pub(crate) fn ncl_service(
+    let pin_iface_name_dir = format!("{folder}/{PIN_IFACE_NAME_FOLDER}");
+    let pin_iface_path = Path::new(&pin_iface_name_dir);
+    if pin_iface_path.exists() {
+        pin_iface_name(&pin_iface_path)?;
```
So the opt-in mechanism here is that the user creates a file?  WDYT about my proposal of a CLI command instead in [the bug](https://bugzilla.redhat.com/show_bug.cgi?id=2182769)?

cathay4t replied:
I prefer `no issue no change` in case there is side impact on this.

cathay4t replied:
Yes. If there is `pin_iface_name` folder in `/etc/nmstate` and `pin.yml` file there, nmstate do the pin work.

cathay4t committed:
DEMO: service: Add support hard coding iface name to MAC

Some OS upgrade might cause interface name changes. To solve that:

 1. Before upgrade,
    * mkdir /etc/nmstate/pin_iface_name/
    * nmstatectl show -k | sudo tee /etc/nmstate/pin_iface_name/pin.yml
    * sudo systemctl enable nmstate

 2. Reboot to new version of OS and let nmstate.service run.
 3. Reboot again.

The `nmstate.service` will create file like if it found any interface
changed name according to stored state in /etc/nmstate/pin_iface_name/pin.yml

    /etc/systemd/network/98-ens9.link

With content:

```
[Match]
MACAddress=01:23:45:67:89:ab

[Link]
Name=ens9
```

After the second reboot, the systemd-udevd will use the old name ens9.

Signed-off-by: Gris Ge <fge@redhat.com>

cgwalters commented:
Having a double reboot here is *possible* but we'll want to be careful here and e.g. ensure that on the next reboot `nmstate-handle-pinning.service` is ordered *before* `NetworkManager.service`; otherwise we may end up doing DHCP on NICs that were supposed to be static, etc.

I was thinking a much simpler flow here is to forcibly pin active NICs unconditionally.  The flow would be:

- running rhel8
- run `nmstatectl pin-nic-names`
- queue upgrade to rhel9
- Optional: run `rpm-ostree initramfs-etc --track /etc/systemd/network` if networking is enabled in the initramfs (I think we can check for `ip=` on kernel cmdline)
- reboot

cgwalters commented:
I did a PR over in https://github.com/nmstate/nmstate/pull/2300 with some non-functional cleanups - want to pull that into this PR?

cathay4t commented:
Closing in the favor of https://github.com/nmstate/nmstate/pull/2301
